### PR TITLE
fix(common): Harden JoinWithFinalSeparator and SelectListHelper edge cases

### LINKED ISSUE #248: Harden pre-existing edge cases in JoinWithFinalSeparator and SelectListHelper
Surfaced by review bots during #245 (analyzer-warning cleanup). Both are **pre-existing** on `master` — not introduced by #245 — and are out of scope for that cleanup PR, so tracking them here.

### 1. `EnumerableExtensions.JoinWithFinalSeparator` — boundary counts
`src/Common/Collections/EnumerableExtensions.cs`

```csharp
return arraySource.Take(count - 1).Join(separator, valueSelector) + finalSeparator + valueSelector(arraySource[arraySource.Length - 1]);
```
- **Empty source:** indexes `arraySource[-1]` → `IndexOutOfRangeException`.
- **Single element:** prepends `finalSeparator` before the only element (e.g. `" and x"` instead of `"x"`).

Add explicit `count == 0` and `count == 1` branches before indexing the last element.

### 2. `SelectListHelper.CreateFor` — null delegate results
`src/Common.WebUI/TagUtilities/SelectListHelper.cs`

`textFunc(item).ToString()` / `valueFunc(item).ToString()` dereference the delegate results without a null check; if a delegate returns `null` for an item, `SelectListItem` construction throws at runtime. Guard with null-conditional/coalescing before building the item.

Both should be covered by unit tests (positive + boundary/negative) per the repo's xUnit v3 testing standards.

### PR DESCRIPTION
**Pull request by kploch**
### Summary

Hardens two pre-existing edge cases surfaced by review bots during #245, per the prescriptions in #248. Both are correctness fixes in shipping code targeted for the v4.0 stable release.

### Changes

- `src/Common/Collections/EnumerableExtensions.cs` — `JoinWithFinalSeparator<TValue, TResult>`:
  - **Empty source** previously indexed `arraySource[-1]` → `IndexOutOfRangeException`; now returns `string.Empty`.
  - **Single element** previously returned `finalSeparator + element` (e.g. `" and x"`); now returns just the element.
  - Implemented as a single `count <= 1` branch that falls back to the existing `Join` overload, so boundary behaviour matches `string.Join` semantics exactly. XML `<returns>` doc updated to document the boundary behaviour.
  - Per review feedback (`1486e0f`): the source is consumed via `IReadOnlyList<TValue>` instead of being copied to an array, avoiding the allocation for arrays/`List<T>`.
- `src/Common.WebUI/TagUtilities/SelectListHelper.cs` — `CreateFor<TModel>`: `textFunc(item).ToString()` / `valueFunc(item).ToString()` threw `NullReferenceException` when a delegate returned `null`; now `?.ToString() ?? string.Empty`, so null results render as empty `Text`/`Value`.
- Regression tests (test-first — each verified failing before the fix):
  - `JoinWithFinalSeparator_should_return_empty_string_for_empty_source` (failed with `IndexOutOfRangeException`)
  - `JoinWithFinalSeparator_should_return_the_only_element_without_separators_for_single_element_source` (failed with `" and only"`)
  - `JoinWithFinalSeparator_should_use_only_final_separator_for_two_element_source` (boundary guard)
  - `CreateFor_should_map_null_delegate_results_to_empty_text_and_value` (failed with `NullReferenceException`)
- `change-log/2026-07-11-248-joinwithfinalseparator-selectlist-edge-cases.md` — change-log entry for the v4.0 release notes.

### Design Decisions

- **`string.Join` boundary semantics** for empty/single sources (empty → `""`, single → bare element) rather than throwing: consistent with the BCL and the natural rendering for human-readable lists. Shipping in v4.0 makes the throw→`""` change acceptable.
- **Null delegate results → `string.Empty`**: `SelectListItem.Text`/`Value` are non-nullable-annotated, and an empty option entry is the neutral rendering.
- **`ToString()` conversion contract kept:** a review suggestion to use `as string` was declined — it would silently discard non-string delegate results (the API deliberately accepts `Func<TModel, object>`).
- **Codex/Gemini plan review skipped (documented deviation):** the Codex CLI rejects all models on the current ChatGPT account and Gemini's free-tier quota is exhausted, so the external second-opinion gate could not run. The fixes are prescribed verbatim in #248 and the PR bot fleet provides independent review.

### Review feedback outcome

All 4 automated review threads triaged, replied to, and resolved:
- Accepted: `IReadOnlyList<TValue>` optimisation (gemini-code-assist) — commit `1486e0f`.
- Declined with reasoning: `as string` conversion change (behavioural regression); full streaming rewrite (out of scope; lookahead complexity vs. short human-readable inputs — partially addressed by the accepted optimisation).
- False positive: codeant-ai's claim that the change-log entry violates the ContextStream rule — `change-log/` is the repo convention (`.claude/rules/change-log.md`) that feeds release-notes generation in `release.yml`.

### Testing

- Full solution: all tests pass (827 in Common.Tests net10.0 / 794 net8.0 incl. 3 new; 3 in WebUI.Tests incl. 1 new), 0 failures, 6 pre-existing unrelated skips.
- Build: zero new warnings (only the pre-existing #250 packaging warning).

## Issue ticket number and link

Closes #248

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] Do we need to implement analytics? — No.
- [x] Will this be part of a product update? If yes, please write one phrase about this update. — Yes: bug fixes — `JoinWithFinalSeparator` handles empty/single-element sources correctly, and `SelectListHelper.CreateFor` no longer throws on null text/value delegate results.

___

## **CodeAnt-AI Description**
Fix edge cases in list joining and select-list item text/value handling

### What Changed
- Joining items with a final separator now works for empty and single-item lists: empty lists return an empty string, and one-item lists return just that item.
- Select-list items now handle null text or value results by showing an empty string instead of failing.
- Added regression tests for empty, single-item, and two-item joins, plus null select-list results.
- Added a release note entry for these fixes.

### Impact
`✅ Fewer crashes on empty lists`
`✅ Correct one-item join output`
`✅ Clearer select-list rendering when values are missing`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: src/Common/Collections/EnumerableExtensions.cs [hunk 1]
```diff
@@ -134,6 +137,11 @@ public static string JoinWithFinalSeparator<TValue, TResult>(this IEnumerable<TV
...
         var count = arraySource.Length;
 
+        if (count <= 1)
+        {
+            return arraySource.Join(separator, valueSelector);
+        }
+
 #pragma warning disable CC0031, IDE0056 // index-from-end operator unavailable on netstandard2.0
         return arraySource.Take(count - 1).Join(separator, valueSelector) + finalSeparator + valueSelector(arraySource[arraySource.Length - 1]); // skipcq: CS-R1019 - index-from-end operator unavailable on netstandard2.0
 #pragma warning restore CC0031, IDE0056
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Collections/EnumerableExtensions.cs]: Accepted — implemented in 1486e0f exactly as suggested: the source is now cast to `IReadOnlyList<TValue>` (falling back to `ToArray()` only for genuinely lazy sequences), avoiding the allocation and copy for arrays, `List<T>` and other read-only list collections. All boundary regression tests pass on both target frameworks.

File: src/Common.WebUI/TagUtilities/SelectListHelper.cs [hunk 2]
```diff
@@ -45,7 +45,7 @@ public static IList<SelectListItem> CreateFor<TModel>(IEnumerable<TModel> items,
 
         // ReSharper disable once PossibleMultipleEnumeration - false-positive
 #pragma warning disable CC0031 // Check for null before calling a delegate - false-positive
-        var result = items.Select(item => new SelectListItem(textFunc(item).ToString(), valueFunc(item).ToString())).ToList();
+        var result = items.Select(item => new SelectListItem(textFunc(item)?.ToString() ?? string.Empty, valueFunc(item)?.ToString() ?? string.Empty)).ToList();
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common.WebUI/TagUtilities/SelectListHelper.cs:48]: Declined — `CreateFor` deliberately accepts `Func<TModel, object>` so callers can map any value (ints, enums, composed objects) and rely on `ToString()` as the documented conversion; the existing tests exercise non-string composition (`m.MyText + m.MyTextSuffix`). The suggested `textFunc(item) as string ?? string.Empty` would silently discard every non-string result (e.g. an int id would render as "" instead of "5") — a behavioural regression, not a hardening. This PR's scope is the null-dereference fix (#248), which `?.ToString() ?? string.Empty` addresses without changing the conversion contract.

File: src/Common/Collections/EnumerableExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Collections/EnumerableExtensions.cs]: Declined for this PR — the materialisation exists because the method must know where the sequence ends to place the final separator, and the streaming rewrite adds a lookahead state machine to a method whose typical inputs are short human-readable lists ("apples, oranges and bananas"). The allocation cost was, however, reduced in 1486e0f (per the sibling gemini-code-assist thread): the source is now used directly via `IReadOnlyList<TValue>` for arrays/lists, so the copy only happens for genuinely lazy sequences — for which a single materialisation is the price of the final-separator semantics. A full streaming implementation would be a perf refactor out of scope for this edge-case fix (#248).

File: change-log/2026-07-11-248-joinwithfinalseparator-selectlist-edge-cases.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+## Fix: Harden `JoinWithFinalSeparator` and `SelectListHelper` edge cases
+
+- **Bug fix** in `Ploch.Common.Collections.EnumerableExtensions.JoinWithFinalSeparator`: an empty source threw `IndexOutOfRangeException` and a single-element source incorrectly prepended the final separator (e.g. `" and x"`). An empty source now returns `string.Empty` and a single-element source returns just that element, matching `string.Join` boundary semantics.
+- **Bug fix** in `Ploch.Common.WebUI.TagUtilities.SelectListHelper.CreateFor`: a `textFunc`/`valueFunc` delegate returning `null` for an item caused a `NullReferenceException`. Null delegate results now map to `string.Empty` for the item's `Text`/`Value`.
+- Added regression tests for the empty, single-element, and two-element `JoinWithFinalSeparator` boundaries and for null `SelectListItem` text/value delegate results.
+
+Refs: #248
```
[the remark this answers is not in the document]
Comment (kploch) [reply change-log/2026-07-11-248-joinwithfinalseparator-selectlist-edge-cases.md:7]: I believe this is a false positive. The custom rule being applied ("don't persist notes/lessons in local files, use ContextStream") targets AI-assistant memory artefacts — lessons, decisions, scratch notes. `change-log/` is not that: it is a deliberate repository convention defined in `.claude/rules/change-log.md` ("When working on a task, and before doing a commit, create a file describing the change"), and the release pipeline (`.github/workflows/release.yml`) generates the GitHub Release notes for each version from these entries, then archives them to `change-log/archive/`. Removing this file would break the v4.0 release-notes generation for this fix. Every prior merged PR with user-visible changes carries the same kind of entry (see e.g. `change-log/2026-07-11-246-takerandom-off-by-one.md` merged in #251).
